_posts/2016-04-25-ryan-de-gruyter-angularjs-components.md

```diff
@@ -0,0 +1,404 @@
+More specifically we call it a component directive or template directive. 
+It is an approach to writing your own custom HTML elements which browsers are able to read and render. 
+HTML comes with a set of pre-defined elements, for example the `<div></div>` element or the `<span></span>` element. 
+
+By combing and nesting these standard HTML tags we can build complex UI widgets. 
+We can change their appearance and behavior dynamically with JavaScript and CSS.
+
+True web components can isolate their structure, appearance and behavior. 
+They make use of a technology called the Shadow DOM, which isolates the component in a separate DOM tree. 
+This element will have its styles and scripts encapsulated, they will not conflict with the styles and scripts inside the parent DOM. 
+
+Angular 2 takes full advantage of this technology, but unfortunately AngularJS, the framework I will be talking about in this article, does not. 
+Components you write and register inside AngularJS do not get isolated into a separate DOM tree. 
+
+Lucky for us we are able to mimic the effect of Web components by using directives. 
+We can write a reusable UI element, declare it with a custom tag and configure it by supplying attributes on the element. 
+My advice is to be sure to use correct naming conventions and a module system so styles and scripts will not conflict with each other.
+
+## Directives and components ##
+
+To create and register a custom element in AngularJS, we can use either methods:
+
+    .directive(name, factoryfunction );
+    .component(name, object);
+
+While components are restricted to custom elements, directives can be used to create both elements as well as custom attributes.
+
+There are 3 types of directives: 
+
+- Component directive
+- Attribute directive
+- Structural directive
+
+A **Component directive** is a directive with a template. 
+The .component() method is a helper method which creates a directive set with default properties.
+
+An **Attribute directive** is declared as an element attribute and they can change the appearance or behavior of an element (ng-change, ng-click, etc...).
+
+A **Structural directive** is an attribute or element that manipulates the DOM by adding or removing DOM elements (ng-if, ng-for, etc...).
+ 
+## When do we use .directive(), and when do we use .component()? ##
+Custom UI elements should be created with the .component() helper method because it:
+
+- enforces best practices and provides optimizations(isolate scope, bindings) 
+- has handy defaults which make it easy to create components.
+- makes migration to Angular 2 easier. 
+- can take advantage of the new component router which will be the default router in Angular 2.
+
+Use the .directive() method when you want to manipulate the DOM by adding or removing elements(Structural directive) or when you want to change the appearance or behavior of an element(Attribute directive).
+
+## Creating a component based AngularJS application ##
```

Comment:
It is better to use the same notation across the whole article (component-based), instead of mixing the two.